packaging/metadata.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import enum
+import typing
+
+from . import specifiers
+from . import utils
+
+if typing.TYPE_CHECKING:
+    from collections.abc import Iterable
+    from typing import List, Optional, Set, Tuple
+
+    from . import requirements
+    from . import version as packaging_version  # Alt name avoids shadowing.
+
+    # Type aliases.
+    _NameAndEmail = Tuple[Optional[str], str]
+    _LabelAndURL = Tuple[str, str]
+
+
+class InvalidMetadata(ValueError):
+    """
+    Invalid metadata found.
+    """
+
+
+@enum.unique
+class MetadataVersion(enum.Enum):
+
+    """
+    Core metadata versions.
+    """
+
+    V1_0 = "1.0"
+    V1_1 = "1.1"
+    V1_2 = "1.2"
+    V2_1 = "2.1"
+    V2_2 = "2.2"
+    v2_3 = "2.3"
```

Comment:
Values on PyPI are "1.0", "1.1", "1.2", "1.3", "2.0", "2.1".

Do we want to be strict here, at the risk of not being able to read certain sets of metadata, or a little more lenient and include values known to exist "in the wild" even if they aren't standardised? It's not as if the value here affects behaviour (at least not in the current code).

Reply:
If we follow through with e.g. `from_email_headers()`, that method (and other `from_*` methods) is the only thing that needs to be able to understand the `2.0` version. In which case we do want to be strict in the enum so that people don't try to generate `2.0` metadata.

Reply:
OK, let me explain where I'm coming from. I recently published `pkg_metadata`, which includes functions to convert the "JSON compatible" form of metadata to and from the "email header" format, as well as a function to read pyproject.toml and convert it to JSON compatible format. These are all (well, apart from the PEP 621 reader) functions I've needed and used frequently in my analysis of PyPI metadata, and one of the key requirements there is to *not* reject invalid data (I need it for analysis of things like "how many projects have used invalid version numbers at some point?"). The `pkg_metadata` library is new, but I've been rewriting these functions over and over for years now, so I'm pretty confident that the design and implementation is pretty solid, for my use case at least 😉 

My hope was that by having these conversion functions, and treating the "JSON compatible dictionary" format as the interchange, we could layer a more application friendly metadata class on top of that, while retaining the ability to handle legacy data. So what was hoping might happen is the following:

1. Adopt the functions from `pkg_metadata`[^1] as top-level functions in the `packaging.metadata` module (explicitly *not* as methods of a `Metadata` class). They work solely with string data and do no validation of input data.
2. Have the `Metadata` class provide `from_dict` and `to_dict` methods that translate between the "JSON compatible dictionary"[^2] format and the class. The class form can provide strict checking of fields like metadata version and the ones that translate to other `packaging` classes.
3. I'm not sure about PEP 621 support - the function in `pkg_metadata` was added as an exercise, and because I have ideas of writing a "toy" build backend, but it doesn't have the real-world use that the other functions do. I'm happy to offer it, and I think that "pyproject to JSON" might be a better API than "pyproject to `Metadata` class". But I don't claim to know what's best in this case.

With this context, I *still* don't know the best answer for metadata version. The JSON-compatible form is required to have `metadata_version` (as well as `name` and `version`) because it's never valid to omit that. So how do we write `to_dict` if we don't record it in the class? Making the caller maintain it independently seems unfriendly. Converting invalid versions to valid ones breaks round-tripping (but "stick to the JSON dict if you want round-tripping" is a valid answer here). Allowing anything risks errors from user-constructed data, which might be an issue (I don't know the use cases for user-constructed data apart from `from_pyproject`, but that method should be setting a valid metadata version anyway).

[^1]: Or at least something with that API.
[^2]: That's a real mouthful. Maybe we need a better term for this - Java has "Plain old Data" classes, maybe Python needs "JSON compatible dictionary"?

Reply:
> one of the key requirements there is to _not_ reject invalid data (I need it for analysis of things like "how many projects have used invalid version numbers at some point?").

Ah, so that's why you want more exact round-tripping!

So are you suggesting just providing validation functions for folks who want to make sure they are not writing out invalid metadata? A key point for this submodule is to make it as easy as possible for tool authors to read and write **correct** metadata. I do get your use-case, though, of just want reformatting functionality.


> treating the "JSON compatible dictionary" format as the interchange

https://github.com/pypa/packaging/issues/383 covers the whole back-and-forth over using dictionaries as the data format versus a class.



> 2\. Have the `Metadata` class provide `from_dict` and `to_dict` methods that translate between the "JSON compatible dictionary"[2](#user-content-fn-2-67335918a5e2f3118bcfdfa56b436670) format and the class.

I'm not sure if the class is still worth it if this submodule has all this code around slinging dictionaries about. At that point you're after validation, and conversion to more rich objects like `Version`. The former can be done as a function, and the latter can be done on-demand. Otherwise what you're proposing is something like `from_email_headers()` -> `Metadata.from_json_dict()` -> `Metadata.to_json_dict()` -> `to_email_headers()` if you're going from `PKG-INFO` to `METADATA`.

Another possibility is to make `Metadata` simply be a structured class that stores (lists of) strings and provides helper methods to get rich objects, similar to what this PR proposes to do for `Name`: there's a `raw_` version and a `canonical_` version of anything where there would be validation of the data. And there can be a `validate` argument to any `from_` and `to_` functions for those instances where you want to make sure everything is compliant.


> The JSON-compatible form is required to have `metadata_version` (as well as `name` and `version`) because it's never valid to omit that. So how do we write `to_dict` if we don't record it in the class?

You can't for that use-case.


> 2\. That's a real mouthful. Maybe we need a better term for this - Java has "Plain old Data" classes, maybe Python needs "JSON compatible dictionary"? [↩](#user-content-fnref-2-67335918a5e2f3118bcfdfa56b436670)

I've seen "JSON dict" before.

Reply:
I do think that both the conflicting needs/usecases mentioned here  (making it easy to enforce correct metadata is generated/provided vs making it possible to handle invalid metadata) are valid usecases but they're not equally important IMO. I think it's OK to carve out round-tripping invalid metadata values in the first pass here and start by rejecting it outright.

There's no reason we have to keep the logic for handling both invalid and valid metadata declarations in the same class even, and solving this for just valid metadata values + formats + declaration has substantial value for both frontends and backend authors. "What is valid metadata values" is fairly well defined, as are the keys for the various bits. That's not the case for invalid metadata.

I see value in the ability to handle invalid metadata, but I don't think that's so important as to be a blocker to having a way to parse and transform valid metadata values in a consistent manner across all tooling with a shared implementation.

FWIW, I do think we should support round-tripping valid metadata content (i.e. let's use lists not sets) except for the TOML, which is much "richer“ in potential formatting choices and can be left to tomlkit and friends.

We should also support round-tripping invalid metadata, but that can wait. It's useful in a smaller set of circumstances and there's no reason we can't add functionality to handle that case later. I'd rather solve the well-defined subset of this space first before trying to solve the invalid metadata case.

> one of the key requirements there is to not reject invalid data (I need it for analysis of things like "how many projects have used invalid version numbers at some point?").

Well... I feel obliged to point out -- the example given does not require being able to represent and store invalid metadata. If we reject metadata that has invalid versions, then you can in fact detect this case based on whether the metadata was rejected due to invalid version values.

Making it clearer *why* metadata is rejected is useful and rich (eg: what field, what format, what underlying exception), not least because it will be possible to make error messages nicer! But again, it's not a blocker for a first iteration of this API, IMO. :)

Reply:
> the example given does not require being able to represent and store invalid metadata

Agreed, but typically I want to report the invalid values, so I can understand *why* people are using invalid values. But I take your point, for that particular case I could just catch the error, report it (with the value) and move on.

My wider use case though is to read all the metadata on PyPI, extract it into a structured format, and then write that format to a database, where I can analyze it further using SQL. That case *does* need a structured format that supports retaining invalid data. I'm happy to accept that's a niche case, though.

> So are you suggesting just providing validation functions for folks who want to make sure they are not writing out invalid metadata? A key point for this submodule is to make it as easy as possible for tool authors to read and write correct metadata. I do get your use-case, though, of just want reformatting functionality.

Not exactly. And I don't think you do completely understand my use case, as it's not just about reformatting. In particular, I dispute that tool authors really need to read "known to be completely correct" data very often.

Unlike the other cases, metadata is a *collection* of values. And consumers typically only care about one or two of those values. Requiring values that the consumer doesn't care about to be valid, just so they can access the values they do want, makes it hard to use the class. For example, consider pip - all we need is the name, version and dependencies. A class that won't even read the data if (say) the metadata version was 2.0 or the description wasn't UTF-8 wouldn't be useful to us, because it wouldn't be acceptable (IMO) for us to fail to install a wheel like that. On the other hand, a program that lists packages might only care that the description was readable, and not give two hoots if the package dependency data was invalid.

For me, this is the important point - it's only *writers* that really care if every metadata field is valid. For readers, they only ever care about a subset of fields.

I'm OK with a class that only writes valid data, and that validates its input. What I'm saying is that I'd rather we didn't lock the APIs that read message-format metadata behind that class - there are *other* valid reasons for wanting to read message-format data, which need to handle invalid values. That's my "bare minimum requirement" here. If the module doesn't provide something equivalent to `pkg_metadata.msg_to_json`, it won't be useful to me. (And I'm not convinced it'll be useful to pip, but that's more debatable).

One further thought about reading message format. If there are multiple errors (description isn't valid UTF-8, metadata version is 2.0, provides_dist isn't a valid requirement) how would those get reported? This is a hard problem, and probably requires exception groups to handle properly. I hit this while trying to use `cattrs` for this, and https://github.com/python-attrs/cattrs/issues/258 is where I reported the problem. A good solution is still a work in progress for them.

> Another possibility is to make `Metadata` simply be a structured class that stores (lists of) strings and provides helper methods to get rich objects, similar to what this PR proposes to do for `Name`: there's a `raw_ version` and a `canonical_` version of anything where there would be validation of the data. And there can be a `validate` argument to any `from_` and `to_` functions for those instances where you want to make sure everything is compliant.

Yes. I don't think the `validate` argument on the `from_` methods would get used much - having the "get the structured value" property raise an exception on invalid data is more fine-grained and useful to consumers (see my above point about consumers only caring about a subset). But will consumers forget that property access could raise an exception? But a validate argument on the `to_` methods sounds reasonable, and should probably even default to `True`. It might also be reasonable to validate on property assignment (outside of the `from_` methods) so that users can never *set* invalid values, even though they can *load* them from source data.

But all of that sounds pretty complex to me, and involves possible use cases that I don't actually know anything about, so treat my opinions as more like guesses here 🙂 I can imagine it would be *very* tricky to get the design right for all use cases if we did this, though.

> https://github.com/pypa/packaging/issues/383 covers the whole back-and-forth over using dictionaries as the data format versus a class.

I'll need to review that, I guess. My opinions have changed as I've worked on my application, so maybe things I was OK with before will no longer seem as obvious to me.

Overall, I'm currently thinking that there isn't a "one size fits all" solution here, and maybe we'll simply have to accept that multiple implementations will remain. I'm a little sad about that, but mostly just because I'll have to keep maintaining `pkg_metadata` and can't merge it in here, which is hardly the end of the world 😉